ocaml/xapi/xapi_pool.ml

```diff
@@ -840,22 +840,27 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
     )
   in
   let assert_sm_features_compatible () =
+    debug
+      "%s Checking whether SM features on the joining host is compatible with \
+       the pool"
+      __FUNCTION__ ;
     (* We consider the case where the existing pool has FOO/m, and the candidate having FOO/n,
        where n >= m, to be compatible. Not vice versa. *)
     let features_compatible coor_features candidate_features =
       (* The pool features must not be reduced or downgraded, although it is fine
          the other way around. *)
       Smint.compat_features coor_features candidate_features = coor_features
     in
-
-    let master_sms = Client.SM.get_all ~rpc ~session_id in
+    let pool_sms = Client.SM.get_all_records ~rpc ~session_id in
     List.iter
-      (fun sm ->
-        let master_sm_type = Client.SM.get_type ~rpc ~session_id ~self:sm in
+      (fun (sm_ref, sm_rec) ->
+        let pool_sm_type = sm_rec.API.sM_type in
+        debug "%s Checking SM %s of name %s in the pool" __FUNCTION__
+          (Ref.string_of sm_ref) sm_rec.sM_name_label ;
         let candidate_sm_ref, candidate_sm_rec =
           match
             Db.SM.get_records_where ~__context
```

Comment:
you have a snapshot of all the records already, you don't need to make any more queries to Db.SM

Reply:
Well this `Db.SM.get_records_where` is run on the candidate/joining host, and it is based on the type of the sm rather than the ref, so I suppose that is Ok. The problematic code is the previous `Client.SM.get...` that is run on the coordinator of the pool to be joined.